rust/kernel/i2c.rs

```diff
@@ -0,0 +1,246 @@
+///
+/// The name of the const is `I2C_DEVICE_ID_TABLE`.
+///
+/// # Examples
+///
+/// ```
+/// use kernel::i2c;
+///
+/// define_i2c_id_table! {u32, [
+///     (i2c::DeviceId(b"test-device1"), Some(0xff)),
+///     (i2c::DeviceId(b"test-device2"), None),
+/// ]};
+/// ```
+#[macro_export]
+macro_rules! define_i2c_id_table {
+    ($data_type:ty, $($t:tt)*) => {
+        $crate::define_id_table!(I2C_DEVICE_ID_TABLE, $crate::i2c::DeviceId, $data_type, $($t)*);
+    }
+}
+
+/// An adapter for the registration of i2c drivers.
+pub struct Adapter<T: Driver>(T);
+
+impl<T: Driver> driver::DriverOps for Adapter<T> {
+    type RegType = bindings::i2c_driver;
+
+    unsafe fn register(
+        reg: *mut Self::RegType,
+        name: &'static CStr,
+        module: &'static ThisModule,
+    ) -> Result {
+        // SAFETY: By the safety requirements of this function (defined in the trait definition),
+        // `reg` is non-null and valid.
+        let i2cdrv = unsafe { &mut *reg };
+
+        i2cdrv.driver.name = name.as_char_ptr();
+        i2cdrv.probe_new = Some(Self::probe_callback);
+        i2cdrv.remove = Some(Self::remove_callback);
+        if let Some(t) = T::I2C_DEVICE_ID_TABLE {
+            i2cdrv.id_table = t.as_ref();
+        }
+
+        // SAFETY:
+        //   - `pdrv` lives at least until the call to `platform_driver_unregister()` returns.
+        //   - `name` pointer has static lifetime.
+        //   - `module.0` lives at least as long as the module.
+        //   - `probe()` and `remove()` are static functions.
+        //   - `of_match_table` is either a raw pointer with static lifetime,
+        //      as guaranteed by the [`driver::IdTable`] type, or null.
+        to_result(unsafe { bindings::i2c_register_driver(module.0, reg) })
+    }
+
+    unsafe fn unregister(reg: *mut Self::RegType) {
+        // SAFETY: By the safety requirements of this function (defined in the trait definition),
+        // `reg` was passed (and updated) by a previous successful call to
+        // `i2c_register_driver`.
+        unsafe { bindings::i2c_del_driver(reg) };
+    }
+}
+
+impl<T: Driver> Adapter<T> {
+    fn get_id_info(client: &Client) -> Option<&'static T::IdInfo> {
+        let table = T::I2C_DEVICE_ID_TABLE?;
+
+        let id = unsafe { bindings::i2c_match_id(table.as_ref(), client.ptr) };
+        if id.is_null() {
+            return None;
+        }
+
+        // SAFETY: `id` is a pointer within the static table, so it's always valid.
+        let offset = unsafe { (*id).driver_data } as *mut core::ffi::c_void;
+        if offset.is_null() {
+            return None;
+        }
+
+        // SAFETY: The offset comes from a previous call to `offset_from` in `IdArray::new`, which
+        // guarantees that the resulting pointer is within the table.
+        let ptr = unsafe {
+            id.cast::<u8>()
+                .offset(offset as _)
```

Comment:
Why are you casting `offset` to a pointer and then almost immediately to an isize?

Reply:
Probably as I only copied that part. Should change that then

Reply:
Now treating it as integer, and checking it for 0. I believe this is what you meant?